larpandoracontent/LArReclustering/ThreeDReclusteringAlgorithm.cc

```diff
@@ -0,0 +1,472 @@
+ *
+ *  @brief  Implementation file for the reclustering algorithm class.
+ *
+ *  $Log: $
+ */
+
+#include "Pandora/AlgorithmHeaders.h"
+#include "Managers/ClusterManager.h"
+
+#include "larpandoracontent/LArHelpers/LArPfoHelper.h"
+#include "larpandoracontent/LArHelpers/LArClusterHelper.h"
+#include "larpandoracontent/LArHelpers/LArPcaHelper.h"
+#include "larpandoracontent/LArHelpers/LArMCParticleHelper.h"
+
+#include "larpandoracontent/LArReclustering/ThreeDReclusteringAlgorithm.h"
+
+using namespace pandora;
+
+namespace lar_content
+{
+
+ThreeDReclusteringAlgorithm::ThreeDReclusteringAlgorithm():
+    m_PfosForReclusteringListName("newShowerParticles3D"),
+    m_hitThresholdForNewPfo(0),
+    m_fomThresholdForReclustering(0.3)
+{
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+ThreeDReclusteringAlgorithm::~ThreeDReclusteringAlgorithm()
+{
+}
+
+//------------------------------------------------------------------------------------------------------------------------------------------
+
+StatusCode ThreeDReclusteringAlgorithm::Run()
+{
+    //Only proceed if successfuly get shower pfo list, and it has at least one pfo
+    const PfoList *pShowerPfoList(nullptr);
+    if (!((STATUS_CODE_SUCCESS == PandoraContentApi::GetList(*this, "ShowerParticles3D", pShowerPfoList)) && pShowerPfoList->size()))
+    {
+        return STATUS_CODE_SUCCESS;
+    }
+
+    m_PfosForReclusteringListName = "newShowerParticles3D";
+    PfoList unchangedPfoList;   
+
+    //Save current pfo list name, so that this can be set as current again at the end of reclustering
+    std::string initialPfosListName;
+    PANDORA_RETURN_RESULT_IF(STATUS_CODE_SUCCESS, !=, PandoraContentApi::GetCurrentListName<Pfo>(*this, initialPfosListName));
+
+    //Some pfos are shower-like and yet include track-like 3D clusters. For the moment I don't want to deal with these.
+    const ClusterList *pShowerClusters(nullptr);
+    PandoraContentApi::GetList(*this, "ShowerClusters3D", pShowerClusters);
+    if(!pShowerClusters) return STATUS_CODE_NOT_FOUND;
+
+    for (const Pfo *const pShowerPfo : *pShowerPfoList)
+    {
+        ClusterList clusterList3D;
+        LArPfoHelper::GetThreeDClusterList(pShowerPfo, clusterList3D);
+       
+        if(!this->PassesCutsForReclustering(pShowerPfo))
+        {
+            unchangedPfoList.push_back(pShowerPfo);
+            continue;
+        }
+
+        //Some pfos are shower-like and yet include track-like 3D clusters. For the moment I don't want to deal with these.
+        if(pShowerClusters->end() == std::find(pShowerClusters->begin(), pShowerClusters->end(), clusterList3D.front())) continue;
+
+        CaloHitList caloHitList3D;
+        clusterList3D.front()->GetOrderedCaloHitList().FillCaloHitList(caloHitList3D);
+
+		//Calculate initial figure of merit 
+        float initialFigureOfMerit=this->GetFigureOfMerit(caloHitList3D);
```

Comment:
Is `initialFigureOfMerit` used anywhere else apart from in the following line? If not, the following line can just be `float minimumFigureOfMerit(this->GetFigureOfMerit(caloHitList3D));`?

Reply:
It can. I've changed it.